base/applications/clipbrd/cliputils.c

```diff
@@ -191,59 +192,146 @@ BOOL IsClipboardFormatSupported(UINT uFormat)
 
         default:
         {
+            if (uFormat == Globals.uCFSTR_FILENAMEA ||
+                uFormat == Globals.uCFSTR_FILENAMEW)
+            {
+                return TRUE;
+            }
+
             return FALSE;
         }
     }
 }
 
-SIZE_T
-GetLineExtentW(
-    IN LPCWSTR lpText,
-    OUT LPCWSTR* lpNextLine)
+LPWSTR WideFromAnsi(LPCSTR pszAnsi)
 {
-    LPCWSTR ptr;
+    INT cchMax = MultiByteToWideChar(CP_ACP, 0, pszAnsi, -1, NULL, 0);
+    LPWSTR pszWide = malloc(cchMax * sizeof(WCHAR));
+    if (!pszWide)
+        return NULL;
+    MultiByteToWideChar(CP_ACP, 0, pszAnsi, -1, pszWide, cchMax);
+    return pszWide;
+}
 
-    /* Find the next line of text (lpText is NULL-terminated) */
-    /* For newlines, focus only on '\n', not on '\r' */
-    ptr = wcschr(lpText, L'\n'); // Find the end of this line.
-    if (ptr)
-    {
-        /* We have the end of this line, go to the next line (ignore the endline in the count) */
-        *lpNextLine = ptr + 1;
-    }
-    else
+LPWSTR GetTextFromClipboard(UINT uFormat, BOOL bOpen)
+{
+    HGLOBAL hGlobal;
+    LPSTR pszA;
+    LPWSTR pszW;
+
+    if (bOpen && !OpenClipboard(Globals.hMainWnd))
+        return NULL;
+
+    hGlobal = GetClipboardData(uFormat);
+    if (hGlobal)
     {
-        /* This line was the last one, go pointing to the terminating NULL */
-        ptr = lpText + wcslen(lpText);
-        *lpNextLine = ptr;
+        switch (uFormat)
+        {
+            case CF_DSPTEXT:
+            case CF_TEXT:
+            case CF_OEMTEXT:
+                pszA = _strdup(GlobalLock(hGlobal));
+                GlobalUnlock(hGlobal);
+                pszW = WideFromAnsi(pszA);
```

Comment:
What I don't like with this method, is that this requires the translation to unicode of the whole text stored in the clipboard, which would cause potential memory exhaustion if what is in the clipboard is a very large text.
The original code was avoiding that by only working on the (limited number of) lines that had to be displayed, and by directly calling either the ansi or the unicode text drawing functions on the specified lines.

Reply:
Okay, I reduced text conversions.